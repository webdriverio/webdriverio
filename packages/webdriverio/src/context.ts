const contextManager = new Map<WebdriverIO.Browser, ContextManager>()

export function getContextManager(browser: WebdriverIO.Browser) {
    const existingContextManager = contextManager.get(browser)
    if (existingContextManager) {
        return existingContextManager
    }

    const newContext = new ContextManager(browser)
    contextManager.set(browser, newContext)
    return newContext
}

/**
 * This class is responsible for managing context in a WebDriver session. Many BiDi commands
 * require to be executed in a specific context. This class is responsible for keeping track
 * of the current context and providing the current context the user is in.
 */
export class ContextManager {
    #browser: WebdriverIO.Browser
    #currentContext?: string

    constructor(browser: WebdriverIO.Browser) {
        this.#browser = browser
        if (!this.#isEnabled()) {
            return
        }

        /**
         * Listens for the 'switchToWindow' browser command to handle context changes.
         * Updates the browsingContext with the context passed in 'switchToWindow'.
         */
        this.#browser.on('command', (event) => {
            /**
             * update frame context if user switches using 'switchToWindow'
             * which is WebDriver Classic only
             */
            if (event.command === 'switchToWindow') {
                this.setCurrentContext(event.body.handle)
            }

            /**
             * reset current context if:
             *   - user uses 'switchToParentFrame' which only impacts WebDriver Classic commands
             *   - user uses 'refresh' which resets the context in Classic and so should in Bidi
             */
            if (
                event.command === 'switchToParentFrame' ||
                event.command === 'refresh'
            ) {
                this.#currentContext = undefined
            }
        })
    }

    /**
     * Only run this session helper if BiDi is enabled and we're not in unit tests.
     */
    #isEnabled () {
        return !process.env.WDIO_UNIT_TESTS && this.#browser.isBidi
    }

    /**
     * set context at the start of the session
     */
    async initialize () {
        if (!this.#isEnabled()) {
            return ''
        }

        const windowHandle = await this.#browser.getWindowHandle()
        this.#currentContext = windowHandle
        return windowHandle
    }

    setCurrentContext (context: string) {
        this.#currentContext = context
    }

    async getCurrentContext () {
        if (!this.#currentContext) {
            return this.initialize()
        }
        return this.#currentContext
    }
}
