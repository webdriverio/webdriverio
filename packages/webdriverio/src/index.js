import WebDriver from 'webdriver'
import { validateConfig, wrapCommand, detectBackend } from '@wdio/config'

import MultiRemote from './multiremote'
import { WDIO_DEFAULTS } from './constants'
import { getPrototype, mobileDetector } from './utils'

/**
 * A method to create a new session with WebdriverIO
 *
 * @param  {Object} [params={}]       Options to create the session with
 * @param  {function} remoteModifier  Modifier function to change the monad object
 * @return {object}                   browser object with sessionId
 */
export const remote = function (params = {}, remoteModifier) {
    const config = validateConfig(WDIO_DEFAULTS, params)
    const modifier = (client, options) => {
        if (typeof remoteModifier === 'function') {
            client = remoteModifier(client, Object.assign(options, config))
        }

        Object.assign(options, config)
        return client
    }

    if (params.user && params.key) {
        params = Object.assign({}, detectBackend(params), params)
    }

    const prototype = getPrototype('browser')

    /**
     * apply mobile check flags to browser scope
     */
    const mobileDetection = mobileDetector(params.capabilities)
    Object.assign(prototype, Object.keys(mobileDetection).reduce((proto, flag) => {
        proto[flag] = { value: mobileDetection[flag] }
        return proto
    }, {}))

    return WebDriver.newSession(params, modifier, prototype, wrapCommand)
}

export const attach = function (params) {
    const prototype = getPrototype('browser')
    return WebDriver.attachToSession(params, null, prototype, wrapCommand)
}

export const multiremote = async function (params = {}) {
    const multibrowser = new MultiRemote()
    const browserNames = Object.keys(params)

    /**
     * create all instance sessions
     */
    await Promise.all(
        browserNames.map((browserName) => {
            validateConfig(WDIO_DEFAULTS, params[browserName])
            const prototype = getPrototype('browser')
            const instance = WebDriver.newSession(params[browserName], null, prototype, wrapCommand)
            return multibrowser.addInstance(browserName, instance)
        })
    )

    /**
     * use attachToSession capability to wrap instances around blank pod
     */
    const prototype = getPrototype('browser')
    const sessionParams = {
        sessionId: '',
        isW3C: multibrowser.instances[browserNames[0]].isW3C
    }
    const driver = WebDriver.attachToSession(sessionParams, ::multibrowser.modifier, prototype, wrapCommand)

    /**
     * in order to get custom command added to multiremote instance we need to pass
     * in the prototype of the multibrowser
     */
    const origAddCommand = ::driver.addCommand
    driver.addCommand = (name, fn) => {
        origAddCommand(name, fn, Object.getPrototypeOf(multibrowser.baseInstance))
    }

    return driver
}
