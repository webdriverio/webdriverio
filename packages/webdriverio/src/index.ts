import logger from '@wdio/logger'

import WebDriver from 'webdriver'
import { DEFAULTS } from 'webdriver'
import { validateConfig } from '@wdio/config'
import { wrapCommand, runFnInFiberContext } from '@wdio/utils'
import { Options, Capabilities } from '@wdio/types'
import type * as WebDriverTypes from 'webdriver'

import MultiRemote from './multiremote'
import type ElementCommands from './commands/element'
import SevereServiceErrorImport from './utils/SevereServiceError'
import detectBackend from './utils/detectBackend'
import { WDIO_DEFAULTS } from './constants'
import {
    getPrototype, addLocatorStrategyHandler, isStub, getAutomationProtocol,
    updateCapabilities
} from './utils'
import type { Browser, MultiRemoteBrowser, AttachOptions } from './types'

export type RemoteOptions = Options.WebdriverIO & Omit<Options.Testrunner, 'capabilities'>

/**
 * A method to create a new session with WebdriverIO
 *
 * @param  {Object} [params={}]       Options to create the session with
 * @param  {function} remoteModifier  Modifier function to change the monad object
 * @return {object}                   browser object with sessionId
 */
export const remote = async function (params: RemoteOptions, remoteModifier?: Function): Promise<Browser<'async'>> {
    logger.setLogLevelsConfig(params.logLevels as any, params.logLevel)

    const config = validateConfig<RemoteOptions>(WDIO_DEFAULTS, params, Object.keys(DEFAULTS) as any)
    const automationProtocol = await getAutomationProtocol(config)
    const modifier = (client: WebDriverTypes.Client, options: Options.WebdriverIO) => {
        /**
         * overwrite instance options with default values of the protocol
         * package (without undefined properties)
         */
        Object.assign(options, Object.entries(config)
            .reduce((a, [k, v]) => (v == null ? a : { ...a, [k]: v }), {}))

        if (typeof remoteModifier === 'function') {
            client = remoteModifier(client, options)
        }

        options.automationProtocol = automationProtocol
        return client
    }

    const prototype = getPrototype('browser')
    const ProtocolDriver = (await import(automationProtocol)).default

    params = Object.assign({}, detectBackend(params), params)
    await updateCapabilities(params, automationProtocol)
    const instance: WebdriverIO.Browser = await ProtocolDriver.newSession(params, modifier, prototype, wrapCommand)

    /**
     * we need to overwrite the original addCommand and overwriteCommand
     * in order to wrap the function within Fibers (only if webdriverio
     * is used with @wdio/cli)
     */
    if ((params as Options.Testrunner).framework && !isStub(automationProtocol)) {
        const origAddCommand = instance.addCommand.bind(instance)
        instance.addCommand = (name: string, fn: Function, attachToElement) => (
            origAddCommand(name, runFnInFiberContext(fn), attachToElement)
        )

        const origOverwriteCommand = instance.overwriteCommand.bind(instance)
        instance.overwriteCommand = (name: string, fn: Function, attachToElement) => (
            origOverwriteCommand<keyof typeof ElementCommands, any, any>(name, runFnInFiberContext(fn), attachToElement)
        )
    }

    instance.addLocatorStrategy = addLocatorStrategyHandler(instance)
    return instance
}

export const attach = async function (params: AttachOptions): Promise<Browser<'async'>> {
    const prototype = getPrototype('browser')

    let automationProtocol = 'webdriver'
    if (params.options?.automationProtocol) {
        automationProtocol = params.options?.automationProtocol
    }
    const ProtocolDriver = (await import(automationProtocol)).default
    return ProtocolDriver.attachToSession(params, undefined, prototype, wrapCommand) as WebdriverIO.Browser
}

export const multiremote = async function (
    params: Capabilities.MultiRemoteCapabilities,
    { automationProtocol }: { automationProtocol?: string } = {}
): Promise<MultiRemoteBrowser<'async'>> {
    const multibrowser = new MultiRemote()
    const browserNames = Object.keys(params)

    /**
     * create all instance sessions
     */
    await Promise.all(
        browserNames.map(async (browserName) => {
            const instance = await remote(params[browserName])
            return multibrowser.addInstance(browserName, instance)
        })
    )

    /**
     * use attachToSession capability to wrap instances around blank pod
     */
    const prototype = getPrototype('browser')
    const sessionParams = isStub(automationProtocol) ? undefined : {
        sessionId: '',
        isW3C: multibrowser.instances[browserNames[0]].isW3C,
        logLevel: multibrowser.instances[browserNames[0]].options.logLevel
    }

    const ProtocolDriver = automationProtocol && isStub(automationProtocol)
        ? require(automationProtocol).default
        : WebDriver
    const driver = ProtocolDriver.attachToSession(
        sessionParams,
        multibrowser.modifier.bind(multibrowser),
        prototype,
        wrapCommand
    ) as WebdriverIO.MultiRemoteBrowser

    /**
     * in order to get custom command overwritten or added to multiremote instance
     * we need to pass in the prototype of the multibrowser
     */
    if (!isStub(automationProtocol)) {
        const origAddCommand = driver.addCommand.bind(driver)
        driver.addCommand = (name: string, fn: Function, attachToElement) => {
            return origAddCommand(
                name,
                runFnInFiberContext(fn),
                attachToElement,
                Object.getPrototypeOf(multibrowser.baseInstance),
                multibrowser.instances
            )
        }

        const origOverwriteCommand = driver.overwriteCommand.bind(driver)
        driver.overwriteCommand = (name: string, fn: Function, attachToElement) => {
            return origOverwriteCommand<keyof typeof ElementCommands, any, any>(
                name,
                runFnInFiberContext(fn),
                attachToElement,
                Object.getPrototypeOf(multibrowser.baseInstance),
                multibrowser.instances
            )
        }
    }

    driver.addLocatorStrategy = addLocatorStrategyHandler(driver)
    return driver
}

export const SevereServiceError = SevereServiceErrorImport
export * from './types'
export * from './utils/interception/types'
