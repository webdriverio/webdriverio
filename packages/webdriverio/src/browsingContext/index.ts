import clone from 'lodash.clonedeep'
import { getBrowserObject } from '@wdio/utils'
import { webdriverMonad } from '@wdio/utils'

import { getContextManager } from '../session/context.js'
import { getPrototype as getWDIOPrototype } from '../utils/index.js'
import type { ContextProps } from './types.js'
import * as pageCommands from '../commands/browsingContext.js'

/**
 * transforms a findElement response into a WDIO element
 * @param  {string} selector  selector that was used to query the element
 * @param  {Object} res       findElement response
 * @return {Object}           WDIO element object
 */
export function getPage(
    this: WebdriverIO.Browser,
    contextId: string,
    props: ContextProps = { isIframe: false, isTab: false, isWindow: false, request: undefined, parent: undefined }
): WebdriverIO.BrowsingContext {
    const browser = getBrowserObject(this)
    const pageCommandKeys = Object.keys(pageCommands)
    const propertiesObject: PropertyDescriptorMap = {
        /**
         * filter out browser commands from object
         */
        ...(Object.entries(clone(browser.__propertiesObject__)).reduce((commands, [name, descriptor]) => {
            if (!pageCommandKeys.includes(name)) {
                commands[name] = descriptor
            }
            return commands
        }, {} as Record<string, PropertyDescriptor>)),
        ...getWDIOPrototype('browsingContext'),
        scope: { value: 'browser' }
    }

    propertiesObject.emit = { value: this.emit.bind(this) }
    const page = webdriverMonad(this.options, (client: WebdriverIO.BrowsingContext) => {
        client.browser = browser
        client.contextId = contextId
        client.isIframe = props.isIframe
        client.isTab = props.isTab
        client.isWindow = props.isWindow
        client.request = props.request
        client.parent = props.parent || browser
        return client
    }, propertiesObject)

    const pageInstance = page(this.sessionId as string)

    const origAddCommand = pageInstance.addCommand.bind(pageInstance)
    pageInstance.addCommand = (name: string, fn: Function) => {
        browser.__propertiesObject__[name] = { value: fn }
        origAddCommand(name, fn)
    }

    return pageInstance
}

export async function getBasePage(this: WebdriverIO.Browser | WebdriverIO.Element) {
    const browser = getBrowserObject(this)
    const contextManager = getContextManager(browser)
    const contextId = await contextManager.getCurrentContext()
    return getPage.call(browser, contextId, { isIframe: false, isTab: false, isWindow: true, request: undefined, parent: undefined })
}
