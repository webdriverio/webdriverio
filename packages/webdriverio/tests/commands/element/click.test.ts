// @ts-ignore mocked (original defined in webdriver package)
import gotMock from 'got'
import { remote } from '../../../src'

const got = gotMock as jest.Mock

describe('click test', () => {
    it('should allow to left click on an element without passing a button type', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click()

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/element/some-elem-123/click')
    })

    it('should allow to left click on an element by passing a button type', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 'left' })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 0 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 0 })

        await elem.click({ button: 0 })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 0 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 0 })
    })

    it('should allow to right click on an element', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 'right' })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].type)
            .toBe('pointerMove')
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 2 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 2 })

        await elem.click({ button: 2 })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].type)
            .toBe('pointerMove')
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 2 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 2 })
    })

    it('should allow to middle click on an element', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 'middle' })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].type)
            .toBe('pointerMove')
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 1 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 1 })

        await elem.click({ button: 1 })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].type)
            .toBe('pointerMove')
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 1 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 1 })
    })

    it('should allow to left click on an element with an offset without passing a button type', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ y: 30 })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].type)
            .toBe('pointerMove')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].x)
            .toBe(0)
        expect(got.mock.calls[2][1].json.actions[0].actions[0].y)
            .toBe(30)
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 0 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 0 })
    })

    it('should allow to right click on an element with an offset and passing a button type', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 2, x: 40, y: 30 })

        expect(got.mock.calls[2][0].pathname)
            .toBe('/session/foobar-123/actions')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].type)
            .toBe('pointerMove')
        expect(got.mock.calls[2][1].json.actions[0].actions[0].x)
            .toBe(40)
        expect(got.mock.calls[2][1].json.actions[0].actions[0].y)
            .toBe(30)
        expect(got.mock.calls[2][1].json.actions[0].actions[1])
            .toStrictEqual({ type: 'pointerDown', button: 2 })
        expect(got.mock.calls[2][1].json.actions[0].actions[2])
            .toStrictEqual({ type: 'pointerUp', button: 2 })
    })

    it('should allow to left click on an element (no w3c)', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar-noW3C'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 'left' })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 25, yoffset: 15 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json).toStrictEqual({ button: 0 })

        await elem.click({ button: 0 })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 25, yoffset: 15 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json).toStrictEqual({ button: 0 })
    })

    it('should allow to right click on an element (no w3c)', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar-noW3C'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 'right' })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 25, yoffset: 15 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json)
            .toStrictEqual({ button: 2 })

        await elem.click({ button: 2 })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 25, yoffset: 15 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json)
            .toStrictEqual({ button: 2 })
    })

    it('should allow to middle click on an element (no w3c)', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar-noW3C'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 'middle' })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 25, yoffset: 15 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json)
            .toStrictEqual({ button: 1 })

        await elem.click({ button: 1 })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 25, yoffset: 15 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json)
            .toStrictEqual({ button: 1 })
    })

    it('should allow to left click on an element with an offset without passing a button type (no w3c)', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar-noW3C'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ y: 30 })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 25, yoffset: 45 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json)
            .toStrictEqual({ button: 0 })
    })

    it('should allow to right click on an element with an offset and passing a button type (no w3c)', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar-noW3C'
            }
        })
        const elem = await browser.$('#foo')

        await elem.click({ button: 2, x: 40, y: 30 })

        expect(got.mock.calls[3][0].pathname)
            .toBe('/session/foobar-123/moveto')
        expect(got.mock.calls[3][1].json)
            .toStrictEqual({ element: 'some-elem-123', xoffset: 65, yoffset: 45 })
        expect(got.mock.calls[4][0].pathname)
            .toBe('/session/foobar-123/click')
        expect(got.mock.calls[4][1].json).toStrictEqual({ button: 2 })
    })

    it('should throw an error if the passed argument is not an options object', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        expect(elem.click([])).rejects.toThrow('Options must be an object')
    })

    it('should throw an error if no valid coördicates are passed', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        expect(elem.click({ x: 'not-suppported' })).rejects.toThrow('Coördinates must be integers')
    })

    it('should throw an error if no valid button type is passed', async () => {
        const browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar'
            }
        })
        const elem = await browser.$('#foo')

        expect(elem.click({ button: 'not-suppported' })).rejects.toThrow('Button type not supported.')
    })

    afterEach(() => {
        got.mockClear()
    })
})
