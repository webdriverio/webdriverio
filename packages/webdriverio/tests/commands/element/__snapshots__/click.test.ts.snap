// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`click test > should allow to left click on an element by passing a button type 1`] = `
{
  "actions": [
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "duration": 100,
      "height": 0,
      "origin": {
        "element-6066-11e4-a52e-4f735466cecf": "some-elem-123",
      },
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerMove",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "height": 0,
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerDown",
      "width": 0,
    },
    {
      "button": 0,
      "type": "pointerUp",
    },
  ],
  "id": "action1",
  "parameters": {
    "pointerType": "mouse",
  },
  "type": "pointer",
}
`;

exports[`click test > should allow to left click on an element by passing a button type 2`] = `
{
  "actions": [
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "duration": 100,
      "height": 0,
      "origin": {
        "element-6066-11e4-a52e-4f735466cecf": "some-elem-123",
      },
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerMove",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "height": 0,
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerDown",
      "width": 0,
    },
    {
      "button": 0,
      "type": "pointerUp",
    },
  ],
  "id": "action2",
  "parameters": {
    "pointerType": "mouse",
  },
  "type": "pointer",
}
`;

exports[`click test > should allow to left click on an element with an offset without passing a button type 1`] = `
[
  {
    "actions": [
      {
        "altitudeAngle": 0,
        "azimuthAngle": 0,
        "button": 0,
        "duration": 100,
        "height": 0,
        "origin": {
          "element-6066-11e4-a52e-4f735466cecf": "some-elem-123",
        },
        "pressure": 0,
        "tangentialPressure": 0,
        "tiltX": 0,
        "tiltY": 0,
        "twist": 0,
        "type": "pointerMove",
        "width": 0,
        "x": 0,
        "y": 15,
      },
      {
        "altitudeAngle": 0,
        "azimuthAngle": 0,
        "button": 0,
        "height": 0,
        "pressure": 0,
        "tangentialPressure": 0,
        "tiltX": 0,
        "tiltY": 0,
        "twist": 0,
        "type": "pointerDown",
        "width": 0,
      },
      {
        "button": 0,
        "type": "pointerUp",
      },
    ],
    "id": "action7",
    "parameters": {
      "pointerType": "mouse",
    },
    "type": "pointer",
  },
]
`;

exports[`click test > should allow to middle click on an element 1`] = `
{
  "actions": [
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "duration": 100,
      "height": 0,
      "origin": {
        "element-6066-11e4-a52e-4f735466cecf": "some-elem-123",
      },
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerMove",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 1,
      "height": 0,
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerDown",
      "width": 0,
    },
    {
      "button": 1,
      "type": "pointerUp",
    },
  ],
  "id": "action5",
  "parameters": {
    "pointerType": "mouse",
  },
  "type": "pointer",
}
`;

exports[`click test > should allow to middle click on an element 2`] = `
{
  "actions": [
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "duration": 100,
      "height": 0,
      "origin": {
        "element-6066-11e4-a52e-4f735466cecf": "some-elem-123",
      },
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerMove",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 1,
      "height": 0,
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerDown",
      "width": 0,
    },
    {
      "button": 1,
      "type": "pointerUp",
    },
  ],
  "id": "action6",
  "parameters": {
    "pointerType": "mouse",
  },
  "type": "pointer",
}
`;

exports[`click test > should allow to right click on an element 1`] = `
{
  "actions": [
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "duration": 100,
      "height": 0,
      "origin": {
        "element-6066-11e4-a52e-4f735466cecf": "some-elem-123",
      },
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerMove",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 2,
      "height": 0,
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerDown",
      "width": 0,
    },
    {
      "button": 2,
      "type": "pointerUp",
    },
  ],
  "id": "action3",
  "parameters": {
    "pointerType": "mouse",
  },
  "type": "pointer",
}
`;

exports[`click test > should allow to right click on an element 2`] = `
{
  "actions": [
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 0,
      "duration": 100,
      "height": 0,
      "origin": {
        "element-6066-11e4-a52e-4f735466cecf": "some-elem-123",
      },
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerMove",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    {
      "altitudeAngle": 0,
      "azimuthAngle": 0,
      "button": 2,
      "height": 0,
      "pressure": 0,
      "tangentialPressure": 0,
      "tiltX": 0,
      "tiltY": 0,
      "twist": 0,
      "type": "pointerDown",
      "width": 0,
    },
    {
      "button": 2,
      "type": "pointerUp",
    },
  ],
  "id": "action4",
  "parameters": {
    "pointerType": "mouse",
  },
  "type": "pointer",
}
`;

exports[`click test > should allow to right click on an element with an offset and passing a button type 1`] = `
{
  "altitudeAngle": 0,
  "azimuthAngle": 0,
  "button": 2,
  "height": 0,
  "pressure": 0,
  "tangentialPressure": 0,
  "tiltX": 0,
  "tiltY": 0,
  "twist": 0,
  "type": "pointerDown",
  "width": 0,
}
`;
